presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/StandardColumnMappings.java

```diff
@@ -446,12 +471,36 @@ public static long toPrestoTimestamp(LocalDateTime localDateTime)
     @Deprecated
     public static LocalDateTime fromPrestoLegacyTimestamp(long value, ZoneId sessionZone)
     {
-        return Instant.ofEpochMilli(value).atZone(sessionZone).toLocalDateTime();
+        return fromPrestoLegacyTimestamp(createTimestampType(3), value, sessionZone);
+    }
+
+    /**
+     * @deprecated applicable in legacy timestamp semantics only
+     */
+    @Deprecated
+    public static LocalDateTime fromPrestoLegacyTimestamp(TimestampType timestampType, long value, ZoneId sessionZone)
+    {
+        return getInstantFromTimestamp(timestampType, value).atZone(sessionZone).toLocalDateTime();
     }
 
     public static LocalDateTime fromPrestoTimestamp(long value)
     {
-        return Instant.ofEpochMilli(value).atZone(UTC).toLocalDateTime();
+        return fromPrestoTimestamp(createTimestampType(3), value);
+    }
+
+    public static LocalDateTime fromPrestoTimestamp(TimestampType timestampType, long value)
+    {
+        return getInstantFromTimestamp(timestampType, value).atZone(UTC).toLocalDateTime();
+    }
+
+    private static Instant getInstantFromTimestamp(TimestampType timestampType, long value)
+    {
+        int precision = timestampType.getPrecision();
+        checkArgument(precision <= MAX_SHORT_PRECISION && precision >= 0, "Precision is out of range");
+        if (precision <= 3) {
+            return Instant.ofEpochMilli(value);
+        }
+        return Timestamps.getInstant(value, 0);
```

Comment:
This is correct, but looking at the code line, it's not obvious.
One has to check parameter names to see long value matches 
the expected representation.